CompuCell3D/core/CompuCell3D/steppables/TubeFieldInitializer/TubeFieldInitializer.cpp

```diff
@@ -0,0 +1,303 @@
+    if (regionVec.size() > 0) {
+        for (int i = 0; i < regionVec.size(); ++i) {
+            TubeFieldInitializerData initData;
+
+            if (regionVec[i]->findElement("Gap"))
+                initData.gap = regionVec[i]->getFirstElement("Gap")->getUInt();
+            if (regionVec[i]->findElement("Width"))
+                initData.width = regionVec[i]->getFirstElement("Width")->getUInt();
+
+            if (!regionVec[i]->getFirstElement("Types"))
+                throw CC3DException(
+                        "TubeInitializer requires Types element inside Region section.See manual for details.");
+            initData.typeNamesString = regionVec[i]->getFirstElement("Types")->getText();
+            parseStringIntoList(initData.typeNamesString, initData.typeNames, ",");
+
+            if (regionVec[i]->findElement("InnerRadius"))
+                initData.innerRadius = regionVec[i]->getFirstElement("InnerRadius")->getUInt();
+            
+            if (regionVec[i]->findElement("OuterRadius"))
+                initData.outerRadius = regionVec[i]->getFirstElement("OuterRadius")->getUInt();
+            
+            if (regionVec[i]->findElement("Extrude")) {
+                CC3DXMLElement * extrudeXML = regionVec[i]->getFirstElement("Extrude");
+                if (extrudeXML->findElement("From") && extrudeXML->findElement("To")) {
+                    CC3DXMLElement * fromPointXML = extrudeXML->getFirstElement("From");
+                    initData.fromPoint.x = fromPointXML->getAttributeAsUInt("x");
+                    initData.fromPoint.y = fromPointXML->getAttributeAsUInt("y");
+                    initData.fromPoint.z = fromPointXML->getAttributeAsUInt("z");
+
+                    CC3DXMLElement * toPointXML = extrudeXML->getFirstElement("To");
+                    initData.toPoint.x = toPointXML->getAttributeAsUInt("x");
+                    initData.toPoint.y = toPointXML->getAttributeAsUInt("y");
+                    initData.toPoint.z = toPointXML->getAttributeAsUInt("z");
+                }
+                else {
+                    throw CC3DException("The 'Extrude' XML tag requires two tags within it: 'From' and 'To'");
+                }
+            }
+            else {
+                throw CC3DException("TubeInitializer requires an XML element named 'Extrude'");
+            }
+
+            initDataVec.push_back(initData);
+        }
+    }
+    else {
+        cerr << "failed if (regionVec.size() > 0)" << endl;
+    }
+}
+
+Dim3D TubeFieldInitializer::getTubeDimensions(const Dim3D &dim, int size) {
+    Dim3D itDim;
+
+    itDim.x = dim.x / size;
+    if (dim.x % size) 
+        itDim.x += 1;
+
+    itDim.y = dim.y / size;
+    if (dim.y % size) 
+        itDim.y += 1;
+
+    itDim.z = dim.z / size;
+    if (dim.z % size) 
+        itDim.z += 1;
+
+    return itDim;
+
+}
+
+double TubeFieldInitializer::distance(double ax, double ay, double az, double bx, double by, double bz) {
```

Comment:
I would use other functions that already have the same functionality. They are defined in 
https://github.com/CompuCell3D/CompuCell3D/blob/master/CompuCell3D/core/Utils/Coordinates3D.h
https://github.com/CompuCell3D/CompuCell3D/blob/master/CompuCell3D/core/PublicUtilities/NumericalUtils.h

or you may try moving addPoints, subtractPoints etc there.I also suggest using operator overloading to implement addition and subtraction:

https://www.geeksforgeeks.org/operator-overloading-cpp/
e.g.
``` cpp
include <iostream> 
using namespace std; 
class Complex { 
private: 
    int real, imag; 
  
public: 
    Complex(int r = 0, int i = 0) 
    { 
        real = r; 
        imag = i; 
    } 
    void print() { cout << real << " + i" << imag << endl; } 
    // The global operator function is made friend of this 
    // class so that it can access private members 
    friend Complex operator+(Complex const& c1, 
                             Complex const& c2); 
}; 
Complex operator+(Complex const& c1, Complex const& c2) 
{ 
    return Complex(c1.real + c2.real, c1.imag + c2.imag); 
} 
int main() 
{ 
    Complex c1(10, 5), c2(2, 4); 
    Complex c3 
        = c1 
          + c2; // An example call to "operator+" 
    c3.print(); 
    return 0; 
}
```

Reply:
Will do. I am working on editing TubeInit so that it can also create wedge-shaped cells. These changes can come in my next PR.